File: cli/gh-teacher/internal/orgpolicy/budget_test.go
package orgpolicy

import "testing"

func TestClassifyBudget_Tiers(t *testing.T) {
	actions := func(amount int, prevent bool) Budget {
		return Budget{
			BudgetScope:         BudgetScopeOrg,
			BudgetProductSKU:    BudgetProductSKUActions,
			BudgetAmount:        amount,
			PreventFurtherUsage: prevent,
		}
	}

	cases := []struct {
		name     string
		budgets  []Budget
		wantTier BudgetTier
	}{
		{"no budgets", nil, BudgetMissing},
		{"unrelated sku only", []Budget{{BudgetScope: BudgetScopeOrg, BudgetProductSKU: "packages", BudgetAmount: 0, PreventFurtherUsage: true}}, BudgetMissing},
		{"wrong scope", []Budget{{BudgetScope: "repository", BudgetProductSKU: BudgetProductSKUActions, BudgetAmount: 0, PreventFurtherUsage: true}}, BudgetMissing},
		{"$0 hard-stop is enforced", []Budget{actions(0, true)}, BudgetEnforced},
		{"$0 alert-only is missing (spend not stopped)", []Budget{actions(0, false)}, BudgetMissing},
		{"$1 hard-stop is ok", []Budget{actions(1, true)}, BudgetOK},
		{"$50 hard-stop is ok (boundary)", []Budget{actions(BudgetWarnThreshold, true)}, BudgetOK},
		{"$50 alert-only is missing", []Budget{actions(BudgetWarnThreshold, false)}, BudgetMissing},
		{"$51 hard-stop warns (boundary)", []Budget{actions(BudgetWarnThreshold+1, true)}, BudgetWarn},
		{"$100 alert-only is missing (stops no spend)", []Budget{actions(100, false)}, BudgetMissing},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := ClassifyBudget(tc.budgets)
			if got.Tier != tc.wantTier {
				t.Errorf("ClassifyBudget tier = %q, want %q", got.Tier, tc.wantTier)
			}
		})
	}
}

func TestClassifyBudget_CarriesAmountAndHardStop(t *testing.T) {
	v := ClassifyBudget([]Budget{{
		BudgetScope:         BudgetScopeOrg,
		BudgetProductSKU:    BudgetProductSKUActions,
		BudgetAmount:        75,
		PreventFurtherUsage: true,
	}})
	if v.Tier != BudgetWarn {
		t.Errorf("tier = %q, want warn", v.Tier)
	}
	if v.Amount != 75 {
		t.Errorf("Amount = %d, want 75", v.Amount)
	}
	if !v.PreventsUsage {
		t.Errorf("PreventsUsage = false, want true")
	}
}

func TestClassifyBudget_FirstMatchWins(t *testing.T) {
	// GitHub allows one budget per scope+SKU; classify the org-scoped Actions
	// budget and ignore other scopes/SKUs in the list.
	budgets := []Budget{
		{BudgetScope: "repository", BudgetProductSKU: BudgetProductSKUActions, BudgetAmount: 999, PreventFurtherUsage: true},
		{BudgetScope: BudgetScopeOrg, BudgetProductSKU: "packages", BudgetAmount: 999, PreventFurtherUsage: true},
		{BudgetScope: BudgetScopeOrg, BudgetProductSKU: BudgetProductSKUActions, BudgetAmount: 0, PreventFurtherUsage: true},
	}
	if got := ClassifyBudget(budgets); got.Tier != BudgetEnforced {
		t.Errorf("tier = %q, want enforced (org+actions $0 hard-stop)", got.Tier)
	}
}

File: web/src/orgPolicy/budget.test.ts
import { describe, expect, it } from "vitest"

import {
  BUDGET_PRODUCT_SKU_ACTIONS,
  BUDGET_SCOPE_ORG,
  BUDGET_WARN_THRESHOLD,
  classifyBudget,
  type Budget,
} from "./budget"

// Web mirror of the CLI's ClassifyBudget tier tests
// (cli/gh-teacher/internal/orgpolicy/budget_test.go). A divergence here is a
// parity bug.

function actions(amount: number, prevent: boolean): Budget {
  return {
    budget_scope: BUDGET_SCOPE_ORG,
    budget_product_sku: BUDGET_PRODUCT_SKU_ACTIONS,
    budget_amount: amount,
    prevent_further_usage: prevent,
  }
}

describe("classifyBudget", () => {
  it.each([
    { name: "no budgets", budgets: [] as Budget[], tier: "missing" },
    {
      name: "unrelated sku only",
      budgets: [
        {
          budget_scope: BUDGET_SCOPE_ORG,
          budget_product_sku: "packages",
          budget_amount: 0,
          prevent_further_usage: true,
        },
      ],
      tier: "missing",
    },
    {
      name: "wrong scope",
      budgets: [
        {
          budget_scope: "repository",
          budget_product_sku: BUDGET_PRODUCT_SKU_ACTIONS,
          budget_amount: 0,
          prevent_further_usage: true,
        },
      ],
      tier: "missing",
    },
    {
      name: "$0 hard-stop enforced",
      budgets: [actions(0, true)],
      tier: "enforced",
    },
    {
      name: "$0 alert-only missing",
      budgets: [actions(0, false)],
      tier: "missing",
    },
    { name: "$1 hard-stop ok", budgets: [actions(1, true)], tier: "ok" },
    {
      name: "$50 hard-stop ok (boundary)",
      budgets: [actions(BUDGET_WARN_THRESHOLD, true)],
      tier: "ok",
    },
    {
      name: "$50 alert-only missing",
      budgets: [actions(BUDGET_WARN_THRESHOLD, false)],
      tier: "missing",
    },
    {
      name: "$51 hard-stop warns (boundary)",
      budgets: [actions(BUDGET_WARN_THRESHOLD + 1, true)],
      tier: "warn",
    },
    {
      name: "$100 alert-only is missing (stops no spend)",
      budgets: [actions(100, false)],
      tier: "missing",
    },
  ])("$name -> $tier", ({ budgets, tier }) => {
    expect(classifyBudget(budgets).tier).toBe(tier)
  })

  it("carries amount and hard-stop flag", () => {
    const v = classifyBudget([actions(75, true)])
    expect(v.tier).toBe("warn")
    expect(v.amount).toBe(75)
    expect(v.preventsUsage).toBe(true)
  })

  it("classifies the org+actions budget, ignoring other scopes/skus", () => {
    const budgets: Budget[] = [
      {
        budget_scope: "repository",
        budget_product_sku: BUDGET_PRODUCT_SKU_ACTIONS,
        budget_amount: 999,
        prevent_further_usage: true,
      },
      {
        budget_scope: BUDGET_SCOPE_ORG,
        budget_product_sku: "packages",
        budget_amount: 999,
        prevent_further_usage: true,
      },
      actions(0, true),
    ]
    expect(classifyBudget(budgets).tier).toBe("enforced")
  })
})
